### PR TITLE
Add cache to auto-import package.json filter

### PR DESCRIPTION
**Pull request by andrewbranch**
I was debugging another issue and noticed that during completions, we go through some work to generate the module specifiers for every export within an ambient module, when obviously that work only needs to be done once per module, not once per export. It made it a pain to debug module specifier generation since I was hitting a breakpoint in one of those functions called with the same inputs over and over, but apparently that work is quite cheap, because I can’t see any appreciable difference locally. Figured I would run the server perf tests though.

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot user test tsserver
@typescript-bot test tsserver top100

**Review by sheetalkamat — APPROVED** [review]

File: src/services/utilities.ts [hunk 1]
```diff
@@ -3663,7 +3663,13 @@ export function createPackageJsonImportFilter(fromFile: SourceFile, preferences:
       ).filter(p => p.parseable);
 
     let usesNodeCoreModules: boolean | undefined;
-    return { allowsImportingAmbientModule, allowsImportingSourceFile, allowsImportingSpecifier };
+    let ambientModuleCache: Map<Symbol, boolean> | undefined;
+    let sourceFileCache: Map<SourceFile, boolean> | undefined;
```
Comment (jakebailey) [inline src/services/utilities.ts:3667]: I don't think we typically use the symbols or nodes themselves as cache keys; all caches I've dealt with or added use symbol/node IDs instead; not sure if that's intentional or not, though.

Comment (sheetalkamat) [reply src/services/utilities.ts:3667]: i think this should be ok now that we are using native maps and life of this cache is well maintained..

Comment (jakebailey) [reply src/services/utilities.ts:3667]: That does make me wonder if we'd gain anything by switching all of the other caches to work directly...

Comment (andrewbranch) [reply src/services/utilities.ts:3667]: I had a vague memory of saying using the object instead of ids is preferred now. But I don’t know if there’s a real difference.

Comment (DanielRosenwasser) [reply src/services/utilities.ts:3667]: Node IDs are set to `0` unconditionally, and set lazily when we need to use a `Node` in a `Map` or `WeakMap`. I tried to remove some usages out of the compiler and didn't find a ton of wins.

https://github.com/microsoft/TypeScript/pull/46595

The biggest risk here is that you keep a `PackageJsonFilter` around across language service operations - leading to either a memory leak, or corrupt state. A `WeakMap` works, but I'd rather we just be a little careful about how we use these.

So I agree with Sheetal - the cache doesn't escape so we should be okay with these - just make sure the `PackageJsonFilter` object doesn't either.

Comment (DanielRosenwasser) [reply src/services/utilities.ts:3667]: Also - these are `Map`s from things to `boolean` - were you trying to anticipate a refactoring later or something? Otherwise, maybe just use `Set`s.

Comment (DanielRosenwasser) [reply src/services/utilities.ts:3667]: Actually, let me rephrase that - using objects/sparse arrays over `Map`s actually is a win, but not if you rely on iteration order.

Comment (andrewbranch) [reply src/services/utilities.ts:3667]: Yeah, this object only lasts a single request.
